### PR TITLE
Show latest commit for file

### PR DESCRIPTION
**Pull request by JakobDev**
If you view a file, you can now see the latest commit that changed that file.

grafik

### CODE REVIEW

**Comment by lunny** [comment]
The space between commit rows and code top bar should be bigger.

**Comment by JakobDev** [comment]
Should we not attach the box, the keep it in sync with the file list?

grafik

**Comment by lunny** [comment]
This is how GH did that.

image

**Comment by JakobDev** [comment]
I know how it looks in GitHub. GitHub uses the same design for a single file and a list of files. We have in Gitea on the file list currently a other design than GitHub, so I the 2 designs should match.

**Comment by yp05327** [comment]
Maybe we can move the `History` button to the same place?
image
image

**Comment by JakobDev** [comment]
> Maybe we can move the `History` button to the same place?

I also thought about this, But I thin this should be part of another PR

File: routers/web/repo/view.go [hunk 1]
```diff
@@ -333,6 +333,35 @@ func renderReadmeFile(ctx *context.Context, subfolder string, readmeFile *git.Tr
 	}
 }
 
+func loadLatestCommitData(ctx *context.Context, latestCommit *git.Commit) bool {
```
Comment (delvh) [inline routers/web/repo/view.go:344]: Is `routers` the correct place for this function?
I don't think so, `services` is probably better.

File: templates/repo/view_file.tmpl [hunk 2]
```diff
@@ -9,6 +9,22 @@
...
+			<div>
+				{{template "repo/latest_commit" .}}
+			</div>
+			<div class="ui text grey right age">
+				{{if .LatestCommit}}
+					{{if .LatestCommit.Committer}}
+						{{TimeSince .LatestCommit.Committer.When ctx.Locale}}
+					{{end}}
+				{{end}}
+			</div>
```
Comment (delvh) [inline templates/repo/view_file.tmpl]: Sounds like the wrong order:
```suggestion
			{{if and (.LatestCommit .LatestCommit.Committer)}}
				<div class="ui text grey right age">
					{{TimeSince .LatestCommit.Committer.When ctx.Locale}}
				</div>
			{{end}}
```

File: templates/repo/latest_commit.tmpl [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+	{{else}}
+		{{if .LatestCommit.Author}}
+			{{ctx.AvatarUtils.AvatarByEmail .LatestCommit.Author.Email .LatestCommit.Author.Name 24 "gt-mr-2"}}
+			<span class="author-wrapper" title="{{.LatestCommit.Author.Name}}"><strong>{{.LatestCommit.Author.Name}}</strong></span>
+		{{end}}
+	{{end}}
+	<a rel="nofollow" class="ui sha label {{if .LatestCommit.Signature}} isSigned {{if .LatestCommitVerification.Verified}} isVerified{{if eq .LatestCommitVerification.TrustStatus "trusted"}}{{else if eq .LatestCommitVerification.TrustStatus "untrusted"}}Untrusted{{else}}Unmatched{{end}}{{else if .LatestCommitVerification.Warning}} isWarning{{end}}{{end}}" href="{{.RepoLink}}/commit/{{PathEscape .LatestCommit.ID.String}}">
+		<span class="shortsha">{{ShortSha .LatestCommit.ID.String}}</span>
+		{{if .LatestCommit.Signature}}
+				{{template "repo/shabox_badge" dict "root" $ "verification" .LatestCommitVerification}}
```
Comment (delvh) [inline templates/repo/latest_commit.tmpl]:
```suggestion
			{{template "repo/shabox_badge" dict "root" $ "verification" .LatestCommitVerification}}
```

File: routers/web/repo/view.go (hunk 1 above)
Comment (JakobDev) [reply routers/web/repo/view.go:344]: This codes mostly sets `ctx.Data`, so it belongs here. It was also the whole time in this file.

File: routers/web/repo/view.go [hunk 4]
```diff
@@ -382,7 +382,7 @@ func renderFile(ctx *context.Context, entry *git.TreeEntry, treeLink, rawLink st
 	ctx.Data["FileName"] = blob.Name()
 	ctx.Data["RawFileLink"] = rawLink + "/" + util.PathEscapeSegments(ctx.Repo.TreePath)
 
-	commit, err := ctx.Repo.Commit.GetCommitByPath(util.PathEscapeSegments(ctx.Repo.TreePath))
+	commit, err := ctx.Repo.Commit.GetCommitByPath(ctx.Repo.TreePath)
```
Comment (delvh) [inline routers/web/repo/view.go]: ?

Comment (JakobDev) [reply routers/web/repo/view.go]: This caused an error on non ASCII paths.

**Review by delvh — APPROVED** [review]

File: templates/repo/latest_commit.tmpl [hunk 5]
```diff
@@ -0,0 +1,31 @@
+{{if not .LatestCommit}}
+	<div class="ui active tiny slow centered inline">…</div>
```
Comment (lunny) [inline templates/repo/latest_commit.tmpl:2]: What's the `...` for?

File: templates/repo/view_file.tmpl [hunk 6]
```diff
@@ -9,6 +9,22 @@
 			<div class="text left gt-whitespace-pre">{{.FileWarning}}</div>
 		</div>
 	{{end}}
+
+	{{if not .ReadmeInList}}
+		<div id="repo-file-commit-box" class="ui top attached header list-header">
```
Comment (denyskon) [inline templates/repo/view_file.tmpl]:
```suggestion
		<div id="repo-file-commit-box" class="ui top attached header list-header gt-mb-4">
```
I agree that this whole UI should be adapted in another PR, but please at least add the margin for now.

**Review by denyskon — APPROVED** [review]

**Comment by lunny** [comment]
Please resolve the conflicts.

**Comment by lunny** [comment]
图片

**Comment by JakobDev** [comment]
Everything is now working again

**Review by denyskon — APPROVED** [review]
